test/test_binary_ufuncs.py

```diff
@@ -1425,31 +1425,34 @@ def test_fmod(self, device, dtype):
         def _reference_implementation(x, mod):
             np_x = x.cpu().numpy()
             np_mod = 0
-            # No type promotion
-            # Issue #47779: https://github.com/pytorch/pytorch/issues/47779
             if torch.is_tensor(mod):
                 np_mod = mod.cpu().numpy()
             else:
                 np_mod = mod
-                # Non XLA platform needs to cast to int
-                if dtype in torch.testing.get_all_int_dtypes() and self.device_type in ['cpu', 'cuda']:
-                    np_mod = int(np_mod)
             exp = np.fmod(np_x, np_mod)
             exp = torch.from_numpy(exp)
 
             res = torch.fmod(x, mod)
-            res = res.to(exp.dtype)
+            common_dtype = torch.promote_types(res.dtype, exp.dtype)
+            exp = exp.to(common_dtype)
+
+            res = res.to(common_dtype)
             self.assertEqual(res, exp)
             # out
-            out = torch.empty(0, device=device, dtype=dtype)
+            out = torch.empty(0, device=device, dtype=common_dtype)
             torch.fmod(x, mod, out=out)
-            out.to(exp.dtype)
+            out.to(common_dtype)
             self.assertEqual(out, exp)
             self.assertEqual(out.size(), torch.Size([10, 10]))
             # in-place
+            if dtype in torch.testing.get_all_int_dtypes():
+                if (torch.is_tensor(mod) and mod.dtype in torch.testing.get_all_fp_dtypes()) or \
+                   type(mod) == float:
+                    self.assertRaises(RuntimeError, lambda: x.fmod_(mod))
+                    return
             x.fmod_(mod)
-            x.to(exp.dtype)
-            self.assertEqual(out, exp)
+            x = x.to(common_dtype)
```

Comment:
This also seems a little wonky since when the operation is performed inplace no type promotion can occur. 

Is part of the challenge with this test that SciPy's type promotion for fmod is divergent? If that's the case this can call assertEqual with `exact_dtype=False`